sigstore/verify/verifier.py

```diff
@@ -133,75 +144,165 @@ def verify(
-        )[0]
-        verify_sct(
-            sct,
-            bundle.signing_certificate,
-            [parent_cert.to_cryptography() for parent_cert in chain],
-            self._trusted_root.ct_keyring(),
-        )
+        # (2): verify the signing certificate's SCT.
+        sct = _get_precertificate_signed_certificate_timestamps(cert)[0]
+        try:
+            verify_sct(
+                sct,
+                cert,
+                [parent_cert.to_cryptography() for parent_cert in chain],
+                self._trusted_root.ct_keyring(),
+            )
+        except VerificationError as e:
+            raise VerificationError(f"failed to verify SCT on signing certificate: {e}")
 
-        # 3) Check that the signing certificate contains the proof claim as the subject
-        # Check usage is "digital signature"
-        usage_ext = bundle.signing_certificate.extensions.get_extension_for_class(
-            KeyUsage
-        )
+        # (3): verify the signing certificate against the Sigstore
+        #      X.509 profile and verify against the given `VerificationPolicy`.
+        usage_ext = cert.extensions.get_extension_for_class(KeyUsage)
         if not usage_ext.value.digital_signature:
             raise VerificationError("Key usage is not of type `digital signature`")
 
-        # Check that extended usage contains "code signing"
-        extended_usage_ext = (
-            bundle.signing_certificate.extensions.get_extension_for_class(
-                ExtendedKeyUsage
-            )
-        )
+        extended_usage_ext = cert.extensions.get_extension_for_class(ExtendedKeyUsage)
         if ExtendedKeyUsageOID.CODE_SIGNING not in extended_usage_ext.value:
             raise VerificationError("Extended usage does not contain `code signing`")
 
-        policy.verify(bundle.signing_certificate)
+        policy.verify(cert)
 
         _logger.debug("Successfully verified signing certificate validity...")
 
-        # 4) Verify that the signature was signed by the public key in the signing certificate
+        # (4): verify the inclusion proof and signed checkpoint for the
+        #      log entry.
+        # (5): verify the inclusion promise for the log entry, if present.
+        entry = bundle.log_entry
+        try:
+            entry._verify(self._trusted_root.rekor_keyring())
+        except VerificationError as exc:
+            raise VerificationError(f"invalid log entry: {exc}")
+
+        # (6): verify that log entry was integrated circa the signing certificate's
+        #      validity period.
+        integrated_time = datetime.fromtimestamp(entry.integrated_time, tz=timezone.utc)
+        if not (
+            bundle.signing_certificate.not_valid_before_utc
+            <= integrated_time
+            <= bundle.signing_certificate.not_valid_after_utc
+        ):
+            raise VerificationError(
+                "invalid signing cert: expired at time of Rekor entry"
+            )
+
+    def verify_dsse(
```

Comment:
NB: This is slightly out of whack with `sign_intoto` in #956, since this operates at the DSSE level while #956 operates at the in-toto level. It might make more sense to turn this into `_verify_dsse` and then expose `verify_intoto` that performs the payload type check as a wrapper on top of `_verify_dsse`.

Reply:
maybe. On the other hand payload handling an application problem with intoto anyway so maybe this is fine?

Reply:
Yeah, `verify_dsse` seems more appropriate to me ultimately (especially since `verify_intoto` would imply that we're actually verifying the in-toto statement, which is not actually happening).

Conversely, any thoughts on turning `sign_intoto` into `sign_dsse`? The main problem I see with that is that it'll rely on the user passing in an opaque `bytes` to sign, which means they'll also need to pass in an appropriate `payload_type` for the PAE encoding. I can see users doing all kinds of potentially incorrect things there, so maybe keeping signing at the JSON payload + in-toto layer is preferable.

Reply:
I think `sign_dsse()` would work as well if the docs are clear about how to use it for the specific intoto use case... but no strong opinions.

Reply:
Sounds good. I'm tempted to make it `sign_dsse` then, but I'll do that with a follow-up PR.